### ISSUE TITLE
Figure out resource limits for users

### ISSUE THREAD

**Issue by celskeggs** [sipb/homeworld#55]
Tier brainstorming:

RAM / COMPUTE / DISK (USABLE)

 * User: 1 GB / 0.15 cores / 4 GB (everyone has this by default -- enough for a couple containers)
 * Developer: 4 GB / 0.6 cores / 32 GB (just submit a form and say something to prove you're not a bot -- enough for six or eight containers)
 * Project: 16 GB / 2.5 cores / 128 GB (provide a project description -- enough for thirty containers)
 * Service: 64 GB / 10 cores / 512 GB (running a service used by a number of people of MIT -- enough for a hundred and something containers)
 * Baby Cthulhu: 256 GB / 20 cores / 2 TB (running a massive service used by most of MIT -- enough for five hundred containers)
 * The Great Old One Himself: 1 TB / 80 cores / 8 TB (holy squid what -- enough for a few thousand containers)

(Based on specs of 256 GB ram / 20 cores / 4 TB disk per server, becoming 2 TB of usable space after RAID or Ceph or whatever.)

A 20 machine cluster would have 25 TB of ram / 2k cores / 4 PB of disk (2 PB after RAID/ceph/whatever), and would be able to support 5 Great Old Ones, 20 Baby Cthulhus, 80 Services, 320 Projects, 1280 Developers, *OR* 5120 Users.

I suspect that these sizes are overly large, but we'll see. It might also be a better idea to not hand out resources on a distinct CPU / RAM / DISK  basis, because then we'll be underutilized -- it might be better to have conversion factors, so that you could trade one for another, but that might lead to other capacity problems.

**Comment by sdukhovni** [comment]
Let users fill out a form to request resources, apply these numbers as quotas based on tier, if we don't have enough capacity left to give a user resources under their quota then we tell them so and go buy more hardware?

**Comment by sdukhovni** [comment]
Although users won't necessarily know how much they need if they're just trying to set up a website for their club or something, so we should also have some kind of quick-startup mode that lets you choose from a list of common use-cases and gives you reasonable resources for those.

**Comment by celskeggs** [comment]
My thinking is as follows:
 * Everyone gets User tier.
 * You can upgrade to Developer tier trivially. "I'm a course 6!" / "I want to try things!" / "I want to develop a project!" (just enough to prevent bot signups)
 * You can easily upgrade to Project -- "I've been writing <X> and want more quota so that I can deploy it" (this involves confirming that they already have code that they've tested with the Developer tier, but otherwise requires nothing.)
 * Upgrading to Service is harder -- "I've been running <project>, and we've run out of quota as our userbase increases. Can we have more quota?" (Probably has a minimum unique user count, like, say, a hundred or two.)
 * Upgrading to Baby Cthulhu is difficult -- "We've been running <project> for a while now, and want to grow to several thousand users, but don't have enough quota. We've tried optimizing our service, and that's helped, but we still need more quota." (Does something useful for the MIT community, has lots of users, has demonstrated that they actually need more quota and it's not just them being wasteful.)
 * Upgrading to The Great Old One Himself is very difficult -- "Literally everyone at MIT uses our service, we've optimized the squid out of it, and done everything we can, and it's still not enough quota and our users are experiencing bad usability because of it." (Literally as important of a project as Hyades itself.)

The reason I don't want to just allow people to select any tier from the get-go (besides Developer) is that any quota we hand out is quota that we can't give to anyone else -- and I don't want to have to buy more hardware for a cluster sitting at 90% idle because we handed out lots of quota that wasn't needed.

**Comment by sdukhovni** [comment]
Yeah, I'm not saying people should select what tier they're on, just that we shouldn't reserve space for people until they specifically request it.

**Comment by sdukhovni** [comment]
So if a person is on Developer tier, we're not going to reserve more than 4G/0.6/32G for them, and if they only request 2G/0.4/16G then we won't reserve any more than that until they ask for more.

**Comment by celskeggs** [comment]
Agreed. Are you including the "user" tier in that? (I consider it reasonable to allocate that tier without any human-level request being necessary.)

**Comment by celskeggs** [comment]
Ahh, yes, that seems reasonable.

**Comment by celskeggs** [comment]
Somewhat more serious naming suggestion: "tiny / small / medium / large / huge / vast"
More boring alternative: "xsmall / small / medium / large / xlarge / xxlarge"

**Comment by dehnert** [comment]
I think the XVM model is "lockers each get 512MB / 50GB; if you need more, ask with justification" -- in some sense, there's two tiers, "default" and "custom". (sql.mit.edu is arguably similar, but quota enforcement mostly doesn't exist.) What's the motivation for having more tiers than that? (Does the code depend on it?) My instinct is that it might make sense to have a split between "user" and "developer" ("automatic" vs. "somebody san checks it", perhaps akin to scripts hostnames today), but beyond that I'd vaguely expect a fairly real discussion that can include figuring out which resources are needed, so I'm not sure that specific tiers are necessary.

(Sorry if the answers to this are obvious to everybody else.)
